overlord/devicestate/handlers.go

```diff
@@ -86,11 +89,17 @@ func sshGenerateKey() (*rsa.PrivateKey, error) {
 	os.Remove(sshKeyFile)
 	os.Remove(sshPublicKeyFile)
 
-	cmd := exec.Command("ssh-keygen", "-t", "rsa", "-b", strconv.FormatUint(keyLength, 10), "-N", "", "-f", sshKeyFile)
-
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return nil, osutil.OutputErr(out, err)
+	switch C.rsa_generate_key(C.uint64_t(keyLength), C.CString(sshKeyFile), C.CString(sshPublicKeyFile)) {
+	case C.RSA_KEY_GENERATION_SEED_FAILURE:
+		return nil, errors.New("Failed to generate RSA key: RNG not seeded.")
+	case C.RSA_KEY_GENERATION_ALLOCATION_FAILURE:
+		return nil, errors.New("Failed to generate RSA key: Could not allocate memory.")
+	case C.RSA_KEY_GENERATION_KEY_GENERATION_FAILURE:
+		return nil, errors.New("Failed to generate RSA key.")
+	case C.RSA_KEY_GENERATION_IO_FAILURE:
+		return nil, errors.New("Failed to generate RSA key: Could not persist keys.")
+	case C.RSA_KEY_GENERATION_SUCCESS:
+		break
```

Comment:
I modelled the enumeration as exhaustive as possible, aiming to capture all possible error conditions. For that, I left out the generic error code.